agenta-cli/agenta/sdk/tracing/tracing_context.py

```diff
@@ -15,14 +16,28 @@ def __init__(self):
 
         ### --- SPANS --- ###
         self.active_span: Optional[CreateSpan] = None
-        self.tracked_spans: Dict[str, CreateSpan] = {}
-        self.closed_spans: List[CreateSpan] = []
+        self.spans: Dict[str, CreateSpan] = {}
+        self.tree = OrderedDict()
 
     def __repr__(self) -> str:
-        return f"TracingContext(trace_id=[{self.trace_id}], active_span=[{self.active_span.id if self.active_span else None}{' ' + self.active_span.spankind if self.active_span else ''}])"
+        return f"TracingContext(trace='{self.trace_id}', spans={[f'{span.id} {span.spankind}' for span in self.spans.values()]}, tree={self.tree})"
 
     def __str__(self) -> str:
         return self.__repr__()
 
+    def push(self, span) -> None:
+        self.active_span = span
+        self.spans[span.id] = span
 
-tracing_context = contextvars.ContextVar(CURRENT_TRACING_CONTEXT_KEY, default=None)
+        if span.parent_span_id is None:
+            ### --- ROOT  SPAN  --- ###
+            self.tree[span.id] = OrderedDict()
+        elif span.parent_span_id in self.tree:
+            ### --- OTHER SPANS --- ###
+            self.tree[span.parent_span_id][span.id] = OrderedDict()
+        else:
+            # ERROR : The parent should have been in the tree.
+            pass
```

Comment:
Is this deliberate?

Reply:
Yes, because tracing should never raise any unhandled Exception. It'd better fail silently than make the user app fail. I'll improve error logging.

Reply:
Ah, got it.